flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java

```diff
@@ -277,27 +280,31 @@ public boolean allocateSlot(
 
     @Override
     public boolean allocateSlot(
-            int index,
+            int requestedIndex,
             JobID jobId,
             AllocationID allocationId,
             ResourceProfile resourceProfile,
             Time slotTimeout) {
         checkRunning();
 
-        Preconditions.checkArgument(index < numberSlots);
+        Preconditions.checkArgument(requestedIndex < numberSlots);
+
+        // The negative requestIndex indicate that the SlotManger allocate a dynamic slot, we
+        // transfer the index to an increasing number not less than the numberSlots.
+        int index = requestedIndex < 0 ? nextDynamicSlotIndex() : requestedIndex;
 
         TaskSlot<T> taskSlot = allocatedSlots.get(allocationId);
         if (taskSlot != null) {
             return isDuplicatedSlot(taskSlot, jobId, resourceProfile, index);
-        } else if (!isIndexAlreadyTaken(index)) {
+        } else if (isIndexAlreadyTaken(index)) {
             LOG.info(
                     "The static slot with index {} is already assigned to another allocation with id {}.",
                     index,
                     taskSlots.get(index).getAllocationId());
             return false;
         }
 
-        resourceProfile = index >= 0 ? defaultSlotResourceProfile : resourceProfile;
+        resourceProfile = index < numberSlots ? defaultSlotResourceProfile : resourceProfile;
```

Comment:
There are 3 occurrences of `index < numberSlots` and 1 occurrence of `index >= numberSlots` in this file.
Let's deduplicate it with a util method.